src/net_processing.cpp

```diff
@@ -3326,12 +3326,15 @@ void ProcessMessage(
         pfrom.fSentAddr = true;
 
         pfrom.vAddrToSend.clear();
-        std::vector<CAddress> vAddr = connman->GetAddresses();
+        std::vector<CAddress> vAddr;
+        if (pfrom.HasPermission(PF_ADDR)) {
+            vAddr = connman->GetAddresses();
+        } else {
+            vAddr = connman->GetAddresses(pfrom.addr.GetNetwork());
+        }
         FastRandomContext insecure_rand;
         for (const CAddress &addr : vAddr) {
-            if (!banman->IsBanned(addr)) {
-                pfrom.PushAddress(addr, insecure_rand);
-            }
+            pfrom.PushAddress(addr, insecure_rand);
```

Comment:
Well this is fascinating. Because of the way we call `PushAddress()` serially with the elements of the vector in order, and each call after `vAddrToSend` is full evicts a random element, then repeatedly sending `GETADDR` will result in `ADDR` responses that leak the entire `m_addr_response_caches[requestor_network].m_addrs_response_cache` _and the order of elements in that vector_. I don't think that's a problem, because the cache is basically public data, but it's an interesting outcome of the way the `vAddrToSend` is constructed.

We should probably limit `CachedAddrResponse.m_addrs_response_cache` to 1000 elements, possibly by adding a max_elements value to the `CAddrMan::GetAddr()` function. Sorry to keep asking you to go deeper into `CAddrMan`!

Reply:
Can we discuss this a bit more?
I don't mind adding that code, I just don't see the motivation. As you point out, this particular leak doesn't matter, because cache is public data. Why limiting to 1000? Maybe another motivation is saving 300 KB of RAM?

Reply:
As someone [points out](https://github.com/bitcoin/bitcoin/issues/19458), there's a slight non-useful inconsistency. Do you think we should address it within this PR?

Reply:
> I just don't see the motivation

As far as I can tell, the intent for this PR was that for each network, we'd create a cache of 1000 addresses, and serve the same list of 1000 addresses (less banned/discouraged) in response to each GETADDR request. Indeed, that's what the functional test is testing for. Instead, we create a cache of 2500, and serve a random 1000 of them to each GETADDR request, with the following distribution:

![image](https://user-images.githubusercontent.com/1063656/86808428-d9a6ff80-c048-11ea-9018-b4c357b83064.png)

(ie the last address record in the cache always appears in the ADDR response, whereas the first 1000 appear with a probability of ~0.22)

That's very unexpected! The fact that neither the author nor any of the reviewers realised that is motivation enough to make the code easier to understand.

Reply:
Right, I guess I'll put together a separate commit just to address 2500/1000 confusion we had even before this PR

Reply:
> I'll put together a separate commit just to address 2500/1000 confusion

That'd be great. I think it can be a separate PR. My suggestion: remove the `ADDRMAN_GETADDR_MAX_PCT` and `ADDRMAN_GETADDR_MAX` constants from addrman and make it the caller's responsibility to set max records/percentage values. Net processing can use 1000/23 and rpc/net can set no max, potentially exposing all address records in addrman to the client.

Reply:
@jnewbery I think that's the right approach.

Reply:
Added the basic fix and updated the test, leaving moving responsibilities for a future PR.